File: backend/routes/serial.js
import express from 'express'
import { v4 as uuidv4 } from 'uuid'
import pool from '../db.js'

const router = express.Router()

// Geração de novo serial (para admin)
router.post('/gerar', async (req, res) => {
  const { nomeEmpresa, email, validadeDias } = req.body
  const codigo = `FARM-${Math.random().toString(36).substring(2, 6).toUpperCase()}-${Math.random().toString(36).substring(2, 6).toUpperCase()}`
  const validade = new Date()
  validade.setDate(validade.getDate() + (validadeDias || 30))

  try {
    await pool.query(`
      INSERT INTO seriais (id, codigo, nome_empresa, email_vinculado, validade_ate, ativo)
      VALUES ($1, $2, $3, $4, $5, true)
    `, [uuidv4(), codigo, nomeEmpresa, email, validade])

    res.json({ status: 'ok', codigo })
  } catch (err) {
    console.error(err)
    res.status(500).json({ status: 'erro', mensagem: 'Erro ao gerar serial' })
  }
})

// Validação de serial no momento de ativação
router.post('/validar', async (req, res) => {
  const { codigo, email } = req.body

  try {
    const result = await pool.query(`
      SELECT * FROM farol_seriais
      WHERE codigo = $1 AND email_vinculado = $2 AND ativo = true AND validade_ate >= NOW() AND farmacia_id IS NULL
    `, [codigo, email])

    if (result.rows.length === 0) {
      return res.status(400).json({ status: 'erro', mensagem: 'Serial inválido, expirado ou já utilizado.' })
    }

    res.json({ status: 'ok', nomeEmpresa: result.rows[0].nome_empresa })
  } catch (err) {
    console.error(err)
    res.status(500).json({ status: 'erro', mensagem: 'Erro ao validar serial' })
  }
})

export default router

File: backend/routers/serial.py
# backend/routers/serial.py

from fastapi import APIRouter
from models import SerialRequest, ValidarSerialRequest
from db import cursor
from uuid import uuid4
from datetime import datetime, timedelta
import random
import string

router = APIRouter()

def gerar_codigo():
    return f"FARM-{''.join(random.choices(string.ascii_uppercase + string.digits, k=4))}-{''.join(random.choices(string.ascii_uppercase + string.digits, k=4))}"

@router.post("/serial/gerar")
def gerar_serial(req: SerialRequest):
    codigo = gerar_codigo()
    validade = datetime.now() + timedelta(days=req.validadeDias)

    cursor.execute("""
        INSERT INTO farol_seriais (id, codigo, nome_empresa, email_vinculado, validade_ate, ativo)
        VALUES (%s, %s, %s, %s, %s, true)
    """, (str(uuid4()), codigo, req.nomeEmpresa, req.email, validade))

    return {"status": "ok", "codigo": codigo}

@router.post("/serial/validar")
def validar_serial(req: ValidarSerialRequest):
    cursor.execute("""
        SELECT * FROM farol_seriais
        WHERE codigo = %s AND email_vinculado = %s AND ativo = true AND validade_ate >= NOW() AND farmacia_id IS NULL
    """, (req.codigo, req.email))

    resultado = cursor.fetchone()
    if not resultado:
        return {"status": "erro", "mensagem": "Serial inválido, expirado ou já utilizado."}

    return {"status": "ok", "nomeEmpresa": resultado[2]}  # nome_empresa
